src/passes/SafeHeap.cpp

```diff
@@ -0,0 +1,316 @@
+        getLoadName(curr),
+        {
+          curr->ptr,
+          builder.makeConst(Literal(int32_t(curr->offset))),
+        },
+        curr->type
+      )
+    );
+  }
+
+  void visitStore(Store* curr) {
+    if (curr->type == unreachable) return;
+    Builder builder(*getModule());
+    replaceCurrent(
+      builder.makeCall(
+        getStoreName(curr),
+        {
+          curr->ptr,
+          builder.makeConst(Literal(int32_t(curr->offset))),
+          curr->value,
+        },
+        none
+      )
+    );
+  }
+};
+
+struct SafeHeap : public Pass {
+  void run(PassRunner* runner, Module* module) override {
+    // add imports
+    addImports(module);
+    // instrument loads and stores
+    PassRunner instrumenter(module);
+    instrumenter.setIsNested(true);
+    instrumenter.add<AccessInstrumenter>();
+    instrumenter.run();
+    // add helper checking funcs and imports
+    addGlobals(module);
+  }
+
+  Name dynamicTopPtr, segfault, alignfault;
+
+  void addImports(Module* module) {
+    // imports
+    dynamicTopPtr = ImportUtils::getImport(*module, ENV, DYNAMICTOP_PTR_IMPORT);
+    if (!dynamicTopPtr.is()) {
+      auto* import = new Import;
+      import->name = dynamicTopPtr = DYNAMICTOP_PTR_IMPORT;
+      import->module = ENV;
+      import->base = DYNAMICTOP_PTR_IMPORT;
+      import->kind = ExternalKind::Global;
+      import->globalType = i32;
+      module->addImport(import);
+    }
+    segfault = ImportUtils::getImport(*module, ENV, SEGFAULT_IMPORT);
+    if (!segfault.is()) {
+      auto* import = new Import;
+      import->name = segfault = SEGFAULT_IMPORT;
+      import->module = ENV;
+      import->base = SEGFAULT_IMPORT;
+      import->kind = ExternalKind::Function;
+      import->functionType = ensureFunctionType("v", module)->name;
+      module->addImport(import);
+    }
+    alignfault = ImportUtils::getImport(*module, ENV, ALIGNFAULT_IMPORT);
+    if (!alignfault.is()) {
+      auto* import = new Import;
+      import->name = alignfault = ALIGNFAULT_IMPORT;
+      import->module = ENV;
+      import->base = ALIGNFAULT_IMPORT;
+      import->kind = ExternalKind::Function;
+      import->functionType = ensureFunctionType("v", module)->name;
+      module->addImport(import);
+    }
+  }
+
+  void addGlobals(Module* module) {
+    // load funcs
+    Load load;
+    for (auto type : { i32, i64, f32, f64 }) {
+      load.type = type;
+      for (Index bytes : { 1, 2, 4, 8 }) {
+        load.bytes = bytes;
+        if (bytes > getWasmTypeSize(load.type)) continue;
+        for (auto signed_ : { true, false }) {
+          load.signed_ = signed_;
+          if (isWasmTypeFloat(load.type) && load.signed_) continue; 
+          for (Index align : { 1, 2, 4, 8 }) {
+            load.align = align;
+            if (align > bytes) continue;
+            for (auto isAtomic : { true, false }) {
+              load.isAtomic = isAtomic;
+              if (load.isAtomic && align != bytes) continue;
```

Comment:
Why not `load.isAtomic && load.align != load.bytes` or `isAtomic && align != bytes`?

Reply:
Good point, I'll make it consistent throughout, using the more concise form.